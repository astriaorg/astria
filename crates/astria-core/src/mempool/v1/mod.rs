pub mod transaction_status;
