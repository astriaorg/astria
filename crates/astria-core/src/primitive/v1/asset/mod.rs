pub mod denom;
pub use denom::{
    Denom,
    IbcPrefixed,
    ParseDenomError,
    TracePrefixed,
};
