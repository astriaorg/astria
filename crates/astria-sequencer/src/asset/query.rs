use anyhow::Context as _;
use astria_core::{
    primitive::v1::asset,
    protocol::{
        abci::AbciErrorCode,
        asset::v1alpha1::AllowedFeeAssetIdsResponse,
    },
};
use cnidarium::Storage;
use prost::Message as _;
use tendermint::abci::{
    request,
    response,
};

use crate::{
    asset::state_ext::StateReadExt as _,
    state_ext::StateReadExt,
};

// Retrieve the full asset denomination given the asset ID.
//
// Example:
// `abci-cli query --path=asset/denom/<DENOM_ID>`
pub(crate) async fn denom_request(
    storage: Storage,
    request: request::Query,
    params: Vec<(String, String)>,
) -> response::Query {
    use astria_core::protocol::asset::v1alpha1::DenomResponse;

    // use the latest snapshot, as this is a lookup of id->denom
    let snapshot = storage.latest_snapshot();
    let asset_id = match preprocess_request(&params) {
        Ok(asset_id) => asset_id,
        Err(err_rsp) => return err_rsp,
    };

    let height = match snapshot.get_block_height().await {
        Ok(height) => height,
        Err(err) => {
            return response::Query {
                code: AbciErrorCode::INTERNAL_ERROR.into(),
                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
                log: format!("failed getting block height: {err:#}"),
                ..response::Query::default()
            };
        }
    };

    let maybe_denom = match snapshot.get_ibc_asset(asset_id).await {
        Ok(maybe_denom) => maybe_denom,
        Err(err) => {
            return response::Query {
                code: AbciErrorCode::INTERNAL_ERROR.into(),
                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
                log: format!("failed to retrieve denomination `{asset_id}`: {err:#}"),
                ..response::Query::default()
            };
        }
    };

    let Some(denom) = maybe_denom else {
        return response::Query {
            code: AbciErrorCode::VALUE_NOT_FOUND.into(),
            info: AbciErrorCode::VALUE_NOT_FOUND.to_string(),
            log: format!("failed to retrieve value for denomination ID`{asset_id}`"),
            ..response::Query::default()
        };
    };

    let payload = DenomResponse {
        height,
        denom,
    }
    .into_raw()
    .encode_to_vec()
    .into();

    let height = tendermint::block::Height::try_from(height).expect("height must fit into an i64");
    response::Query {
        code: tendermint::abci::Code::Ok,
        key: request.path.into_bytes().into(),
        value: payload,
        height,
        ..response::Query::default()
    }
}

fn preprocess_request(params: &[(String, String)]) -> anyhow::Result<asset::Id, response::Query> {
    let Some(asset_id) = params.iter().find_map(|(k, v)| (k == "id").then_some(v)) else {
        return Err(response::Query {
            code: AbciErrorCode::INVALID_PARAMETER.into(),
            info: AbciErrorCode::INVALID_PARAMETER.to_string(),
            log: "path did not contain asset ID parameter".into(),
            ..response::Query::default()
        });
    };
    let asset_id = hex::decode(asset_id)
        .context("failed decoding hex encoded bytes")
        .and_then(|addr| {
            asset::Id::try_from_slice(&addr).context("failed constructing asset ID from bytes")
        })
        .map_err(|err| response::Query {
            code: AbciErrorCode::INVALID_PARAMETER.into(),
            info: AbciErrorCode::INVALID_PARAMETER.to_string(),
            log: format!("asset ID could not be constructed from provided parameter: {err:#}"),
            ..response::Query::default()
        })?;
    Ok(asset_id)
}

pub(crate) async fn allowed_fee_asset_ids_request(
    storage: Storage,
    request: request::Query,
    _params: Vec<(String, String)>,
) -> response::Query {
    // get last snapshot
    let snapshot = storage.latest_snapshot();

    // get height from snapshot
    let height = match snapshot.get_block_height().await {
        Ok(height) => height,
        Err(err) => {
            return response::Query {
                code: AbciErrorCode::INTERNAL_ERROR.into(),
                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
                log: format!("failed getting block height: {err:#}"),
                ..response::Query::default()
            };
        }
    };

    // get ids from snapshot at height
    let fee_asset_ids = match snapshot.get_allowed_fee_assets().await {
        Ok(maybe_fee_assets) => maybe_fee_assets,
        Err(err) => {
            return response::Query {
                code: AbciErrorCode::INTERNAL_ERROR.into(),
                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
                log: format!("failed to retrieve allowed fee assets: {err:#}"),
                ..response::Query::default()
            };
        }
    };

    let payload = AllowedFeeAssetIdsResponse {
        height,
        fee_asset_ids,
    }
    .into_raw()
    .encode_to_vec()
    .into();

    let height = tendermint::block::Height::try_from(height).expect("height must fit into an i64");
    response::Query {
        code: tendermint::abci::Code::Ok,
        key: request.path.into_bytes().into(),
        value: payload,
        height,
        ..response::Query::default()
    }
}
