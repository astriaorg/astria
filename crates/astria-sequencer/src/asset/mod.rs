use std::sync::OnceLock;

use proto::native::sequencer::v1alpha1::asset::Denom;

pub(crate) static NATIVE_ASSET: OnceLock<Denom> = OnceLock::new();

pub(crate) fn initialize_native_asset(native_asset: &str) {
    if NATIVE_ASSET.get().is_some() {
        tracing::error!("native asset should only be set once");
        return;
    }

    let denom = Denom::from_base_denom(native_asset);
    NATIVE_ASSET
        .set(denom)
        .expect("native asset should only be set once");
}

pub(crate) fn get_native_asset() -> &'static Denom {
    NATIVE_ASSET.get().expect("native asset should be set")
}
