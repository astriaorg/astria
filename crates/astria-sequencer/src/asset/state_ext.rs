use anyhow::{
    bail,
    Context as _,
    Result,
};
use astria_core::primitive::v1::{
    asset,
    asset::Denom,
};
use async_trait::async_trait;
use borsh::{
    BorshDeserialize,
    BorshSerialize,
};
use cnidarium::{
    StateRead,
    StateWrite,
};
use hex::ToHex as _;
use tracing::instrument;

/// Newtype wrapper to read and write a denomination trace from rocksdb.
#[derive(BorshSerialize, BorshDeserialize, Debug)]
struct DenominationTrace(String);

fn asset_storage_key(asset: asset::Id) -> String {
    format!("asset/{}", asset.encode_hex::<String>())
}

#[async_trait]
pub(crate) trait StateReadExt: StateRead {
    #[instrument(skip(self))]
    async fn has_ibc_asset(&self, id: asset::Id) -> Result<bool> {
        match self
            .get_raw(&asset_storage_key(id))
            .await
            .context("failed reading raw asset from state")?
        {
            Some(_) => Ok(true),
            None => Ok(false),
        }
    }

    #[instrument(skip(self))]
    async fn get_ibc_asset(&self, id: asset::Id) -> Result<Denom> {
        let Some(bytes) = self
            .get_raw(&asset_storage_key(id))
            .await
            .context("failed reading raw asset from state")?
        else {
            bail!("asset not found");
        };

        let DenominationTrace(denom_str) =
            DenominationTrace::try_from_slice(&bytes).context("invalid asset bytes")?;
        let denom: Denom = denom_str.into();
        Ok(denom)
    }
}

impl<T: ?Sized + StateRead> StateReadExt for T {}

#[async_trait]
pub(crate) trait StateWriteExt: StateWrite {
    #[instrument(skip(self))]
    fn put_ibc_asset(&mut self, id: asset::Id, asset: &Denom) -> Result<()> {
        let bytes = borsh::to_vec(&DenominationTrace(asset.denomination_trace()))
            .context("failed to serialize asset")?;
        self.put_raw(asset_storage_key(id), bytes);
        Ok(())
    }
}

impl<T: StateWrite> StateWriteExt for T {}

#[cfg(test)]
mod test {
    use astria_core::primitive::v1::asset::{
        Denom,
        Id,
    };
    use cnidarium::StateDelta;

    use super::{
        StateReadExt as _,
        StateWriteExt as _,
    };

    #[tokio::test]
    async fn get_ibc_asset_non_existent() {
        let storage = cnidarium::TempStorage::new().await.unwrap();
        let snapshot = storage.latest_snapshot();
        let state = StateDelta::new(snapshot);

        let asset = Id::from_denom("asset");

        // gets for non existing assets fail
        state
            .get_ibc_asset(asset)
            .await
            .expect_err("gets for non existing ibc assets should fail");
    }

    #[tokio::test]
    async fn has_ibc_asset() {
        let storage = cnidarium::TempStorage::new().await.unwrap();
        let snapshot = storage.latest_snapshot();
        let mut state = StateDelta::new(snapshot);

        let denom = Denom::from_base_denom("asset");

        // non existing calls are ok for 'has'
        assert!(
            !state
                .has_ibc_asset(denom.id())
                .await
                .expect("'has' for non existing ibc assets should be ok"),
            "query for non existing asset should return false"
        );

        state
            .put_ibc_asset(denom.id(), &denom)
            .expect("putting ibc asset should not fail");

        // existing calls are ok for 'has'
        assert!(
            state
                .has_ibc_asset(denom.id())
                .await
                .expect("'has' for existing ibc assets should be ok"),
            "query for existing asset should return true"
        );
    }

    #[tokio::test]
    async fn put_ibc_asset_simple() {
        let storage = cnidarium::TempStorage::new().await.unwrap();
        let snapshot = storage.latest_snapshot();
        let mut state = StateDelta::new(snapshot);

        // can write new
        let denom = Denom::from_base_denom("asset");
        state
            .put_ibc_asset(denom.id(), &denom)
            .expect("putting ibc asset should not fail");
        assert_eq!(
            state
                .get_ibc_asset(denom.id())
                .await
                .expect("an ibc asset was written and must exist inside the database"),
            denom,
            "stored ibc asset was not what was expected"
        );
    }

    #[tokio::test]
    async fn put_ibc_asset_complex() {
        let storage = cnidarium::TempStorage::new().await.unwrap();
        let snapshot = storage.latest_snapshot();
        let mut state = StateDelta::new(snapshot);

        // can write new
        let denom = Denom::from_base_denom("asset_0");
        state
            .put_ibc_asset(denom.id(), &denom)
            .expect("putting ibc asset should not fail");
        assert_eq!(
            state
                .get_ibc_asset(denom.id())
                .await
                .expect("an ibc asset was written and must exist inside the database"),
            denom,
            "stored ibc asset was not what was expected"
        );

        // can write another without affecting original
        let denom_1 = Denom::from_base_denom("asset_1");
        state
            .put_ibc_asset(denom_1.id(), &denom_1)
            .expect("putting ibc asset should not fail");
        assert_eq!(
            state
                .get_ibc_asset(denom_1.id())
                .await
                .expect("an additional ibc asset was written and must exist inside the database"),
            denom_1,
            "additional ibc asset was not what was expected"
        );
        assert_eq!(
            state
                .get_ibc_asset(denom.id())
                .await
                .expect("an ibc asset was written and must exist inside the database"),
            denom,
            "original ibc asset was not what was expected"
        );
    }

    #[tokio::test]
    async fn put_ibc_asset_can_write_unrelated_ids_to_denoms() {
        let storage = cnidarium::TempStorage::new().await.unwrap();
        let snapshot = storage.latest_snapshot();
        let mut state = StateDelta::new(snapshot);

        // can write unrelated ids and denoms
        let id_key = Id::from_denom("asset_0");
        let denom = Denom::from_base_denom("asset_1");
        state
            .put_ibc_asset(id_key, &denom)
            .expect("putting ibc asset should not fail");

        // see that id key and denom's stored id differ
        assert_ne!(
            state
                .get_ibc_asset(id_key)
                .await
                .expect("an ibc asset was written and must exist inside the database")
                .id(),
            id_key,
            "stored ibc asset was not what was expected"
        );
    }
}
