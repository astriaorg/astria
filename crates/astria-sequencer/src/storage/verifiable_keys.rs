use std::{
    fmt::{
        self,
        Display,
        Formatter,
    },
    str::FromStr,
};

use astria_core::primitive::v1::asset::IbcPrefixed;
use astria_eyre::eyre::{
    ContextCompat as _,
    Result,
    WrapErr as _,
};

use crate::accounts::AddressBytes;

pub(crate) mod accounts {
    use super::*;

    pub(crate) const TRANSFER_BASE_FEE_KEY: &str = "transferfee";
    const ACCOUNTS_PREFIX: &str = "accounts/";
    const BALANCE_PREFIX: &str = "balance/";
    const NONCE: &str = "nonce";

    /// Example: `accounts/0101....0101/balance/`.
    ///                   |40 hex chars|
    pub(crate) fn balance_prefix<TAddress: AddressBytes>(address: &TAddress) -> String {
        format!(
            "{}/{BALANCE_PREFIX}",
            AddressPrefixer::new(ACCOUNTS_PREFIX, address)
        )
    }

    /// Example: `accounts/0101....0101/balance/0202....0202`.
    ///                   |40 hex chars|       |64 hex chars|
    pub(crate) fn balance_key<TAddress: AddressBytes, TAsset: Into<IbcPrefixed>>(
        address: TAddress,
        asset: TAsset,
    ) -> String {
        format!(
            "{}/{BALANCE_PREFIX}{}",
            AddressPrefixer::new(ACCOUNTS_PREFIX, &address),
            Asset::from(asset)
        )
    }

    /// Example: `accounts/0101....0101/nonce`.
    ///                   |40 hex chars|
    pub(crate) fn nonce_key<TAddress: AddressBytes>(address: &TAddress) -> String {
        format!("{}/{NONCE}", AddressPrefixer::new(ACCOUNTS_PREFIX, address))
    }

    pub(crate) fn extract_asset_from_key(key: &str) -> Result<IbcPrefixed> {
        Ok(key
            .strip_prefix(ACCOUNTS_PREFIX)
            .and_then(|s| s.split_once(BALANCE_PREFIX).map(|(_, asset)| asset))
            .wrap_err("failed to strip prefix from account balance key")?
            .parse::<Asset>()
            .wrap_err("failed to parse storage key suffix as address hunk")?
            .0)
    }

    #[cfg(test)]
    mod tests {
        use astria_core::primitive::v1::{
            asset::Denom,
            Address,
        };

        use super::*;

        #[test]
        fn keys_should_not_change() {
            let address: Address = "astria1rsxyjrcm255ds9euthjx6yc3vrjt9sxrm9cfgm"
                .parse()
                .unwrap();
            let asset = "an/asset/with/a/prefix".parse::<Denom>().unwrap();
            assert_eq!(
                balance_key(address, &asset),
                balance_key(address, asset.to_ibc_prefixed())
            );
            insta::assert_snapshot!(
                balance_prefix(&address),
                @"accounts/1c0c490f1b5528d8173c5de46d131160e4b2c0c3/balance/"
            );
            insta::assert_snapshot!(
                balance_key(address, asset),
                @"accounts/1c0c490f1b5528d8173c5de46d131160e4b2c0c3/balance/be429a02d00837245167a26\
                16674a979a2ac6f9806468b48a975b156ad711320"
            );
            insta::assert_snapshot!(
                nonce_key(&address),
                @"accounts/1c0c490f1b5528d8173c5de46d131160e4b2c0c3/nonce"
            );
        }

        #[test]
        fn balance_prefix_should_be_prefix_of_balance_key() {
            let address: Address = "astria1rsxyjrcm255ds9euthjx6yc3vrjt9sxrm9cfgm"
                .parse()
                .unwrap();
            let asset = "an/asset/with/a/prefix".parse::<Denom>().unwrap();
            let prefix = balance_prefix(&address);
            let key = balance_key(address, asset);
            assert!(key.strip_prefix(&prefix).is_some());
        }

        #[test]
        fn should_extract_asset_from_key() {
            let asset = IbcPrefixed::new([2; 32]);
            let key = balance_key([1; 20], asset);
            let recovered_asset = extract_asset_from_key(&key).unwrap();
            assert_eq!(asset, recovered_asset);
        }
    }
}

pub(crate) mod address {
    pub(crate) const BASE_PREFIX_KEY: &str = "prefixes/base";
    pub(crate) const IBC_COMPAT_PREFIX_KEY: &str = "prefixes/ibc-compat";
}

pub(crate) mod app {
    pub(crate) const CHAIN_ID_KEY: &str = "chain_id";
    pub(crate) const REVISION_NUMBER_KEY: &str = "revision_number";
    pub(crate) const BLOCK_HEIGHT_KEY: &str = "block_height";
    pub(crate) const BLOCK_TIMESTAMP_KEY: &str = "block_timestamp";
}

pub(crate) mod assets {
    use astria_core::primitive::v1::asset::IbcPrefixed;

    use crate::storage::verifiable_keys::Asset;

    pub(crate) const ASSET_PREFIX: &str = "asset/";
    pub(crate) const NATIVE_ASSET_KEY: &str = "nativeasset";

    /// Example: `asset/0101....0101`.
    ///                |64 hex chars|
    pub(crate) fn asset_key<TAsset: Into<IbcPrefixed>>(asset: TAsset) -> String {
        format!("{ASSET_PREFIX}{}", Asset::from(asset))
    }

    #[cfg(test)]
    mod tests {
        use astria_core::primitive::v1::asset::Denom;

        use super::*;

        #[test]
        fn keys_should_not_change() {
            let asset = "an/asset/with/a/prefix".parse::<Denom>().unwrap();
            insta::assert_snapshot!(
                asset_key(asset),
                @"asset/be429a02d00837245167a2616674a979a2ac6f9806468b48a975b156ad711320"
            );
        }
    }
}

pub(crate) mod authority {
    pub(crate) const SUDO_STORAGE_KEY: &str = "sudo";
    pub(crate) const VALIDATOR_SET_STORAGE_KEY: &str = "valset";
}

pub(crate) mod bridge {
    use crate::{
        accounts::AddressBytes,
        storage::verifiable_keys::AddressPrefixer,
    };

    pub(crate) const INIT_BRIDGE_ACCOUNT_BASE_FEE_KEY: &str = "initbridgeaccfee";
    pub(crate) const BRIDGE_LOCK_BYTE_COST_MULTIPLIER_KEY: &str = "bridgelockmultiplier";
    pub(crate) const BRIDGE_SUDO_CHANGE_FEE_KEY: &str = "bridgesudofee";

    pub(in crate::storage) const BRIDGE_ACCOUNT_PREFIX: &str = "bridgeacc/";
    const BRIDGE_ACCOUNT_SUDO_PREFIX: &str = "bsudo/";
    const BRIDGE_ACCOUNT_WITHDRAWER_PREFIX: &str = "bwithdrawer/";

    /// Example: `bridgeacc/0101....0101/rollupid`.
    ///                    |40 hex chars|
    pub(crate) fn rollup_id_key<T: AddressBytes>(address: &T) -> String {
        format!(
            "{}/rollupid",
            AddressPrefixer::new(BRIDGE_ACCOUNT_PREFIX, address)
        )
    }

    /// Example: `bridgeacc/0101....0101/assetid`.
    ///                    |40 hex chars|
    pub(crate) fn asset_id_key<T: AddressBytes>(address: &T) -> String {
        format!(
            "{}/assetid",
            AddressPrefixer::new(BRIDGE_ACCOUNT_PREFIX, address)
        )
    }

    /// Example: `bsudo/0101....0101`.
    ///                |40 hex chars|
    pub(crate) fn bridge_account_sudo_address_key<T: AddressBytes>(address: &T) -> String {
        AddressPrefixer::new(BRIDGE_ACCOUNT_SUDO_PREFIX, address).to_string()
    }

    /// Example: `bwithdrawer/0101....0101`.
    ///                      |40 hex chars|
    pub(crate) fn bridge_account_withdrawer_address_key<T: AddressBytes>(address: &T) -> String {
        AddressPrefixer::new(BRIDGE_ACCOUNT_WITHDRAWER_PREFIX, address).to_string()
    }

    /// Example: `bridgeacc/0101....0101/withdrawalevent/<event id>`.
    ///                    |40 hex chars|               |UTF-8 chars|
    pub(crate) fn bridge_account_withdrawal_event_key<T: AddressBytes>(
        address: &T,
        withdrawal_event_id: &str,
    ) -> String {
        format!(
            "{}/withdrawalevent/{}",
            AddressPrefixer::new(BRIDGE_ACCOUNT_PREFIX, address),
            withdrawal_event_id
        )
    }

    #[cfg(test)]
    mod tests {
        use astria_core::primitive::v1::Address;

        use super::*;

        #[test]
        fn keys_should_not_change() {
            let address: Address = "astria1rsxyjrcm255ds9euthjx6yc3vrjt9sxrm9cfgm"
                .parse()
                .unwrap();

            insta::assert_snapshot!(
                rollup_id_key(&address),
                @"bridgeacc/1c0c490f1b5528d8173c5de46d131160e4b2c0c3/rollupid"
            );
            insta::assert_snapshot!(
                asset_id_key(&address),
                @"bridgeacc/1c0c490f1b5528d8173c5de46d131160e4b2c0c3/assetid"
            );
            insta::assert_snapshot!(
                bridge_account_sudo_address_key(&address),
                @"bsudo/1c0c490f1b5528d8173c5de46d131160e4b2c0c3"
            );
            insta::assert_snapshot!(
                bridge_account_withdrawer_address_key(&address),
                @"bwithdrawer/1c0c490f1b5528d8173c5de46d131160e4b2c0c3"
            );
            insta::assert_snapshot!(
                bridge_account_withdrawal_event_key(&address, "the-event"),
                @"bridgeacc/1c0c490f1b5528d8173c5de46d131160e4b2c0c3/withdrawalevent/the-event"
            );
        }
    }
}

pub(crate) mod ibc {
    use astria_core::primitive::v1::asset::IbcPrefixed;
    use ibc_types::core::channel::ChannelId;

    use crate::{
        accounts::AddressBytes,
        storage::verifiable_keys::{
            AddressPrefixer,
            Asset,
        },
    };

    pub(crate) const IBC_SUDO_KEY: &str = "ibcsudo";
    pub(crate) const ICS20_WITHDRAWAL_BASE_FEE_KEY: &str = "ics20withdrawalfee";

    const IBC_RELAYER_PREFIX: &str = "ibc-relayer/";

    /// Example: `ibc-data/channel-xxx/balance/0101....0101`.
    ///                           |int|       |64 hex chars|
    pub(crate) fn channel_balance_key<TAsset: Into<IbcPrefixed>>(
        channel: &ChannelId,
        asset: TAsset,
    ) -> String {
        format!("ibc-data/{channel}/balance/{}", Asset::from(asset),)
    }

    /// Example: `ibc-relayer/0101....0101`.
    ///                      |40 hex chars|
    pub(crate) fn ibc_relayer_key<T: AddressBytes>(address: &T) -> String {
        AddressPrefixer::new(IBC_RELAYER_PREFIX, address).to_string()
    }

    #[cfg(test)]
    mod tests {
        use astria_core::primitive::v1::Address;

        use super::*;

        #[test]
        fn keys_should_not_change() {
            let channel = ChannelId::new(5);
            let asset = "an/asset/with/a/prefix"
                .parse::<astria_core::primitive::v1::asset::Denom>()
                .unwrap();
            assert_eq!(
                channel_balance_key(&channel, &asset),
                channel_balance_key(&channel, asset.to_ibc_prefixed()),
            );
            insta::assert_snapshot!(
                channel_balance_key(&channel, &asset),
                @"ibc-data/channel-5/balance/be429a02d00837245167a2616674a979a2ac6f9806468b48a975b1\
                56ad711320"
            );

            let address: Address = "astria1rsxyjrcm255ds9euthjx6yc3vrjt9sxrm9cfgm"
                .parse()
                .unwrap();
            insta::assert_snapshot!(
                ibc_relayer_key(&address),
                @"ibc-relayer/1c0c490f1b5528d8173c5de46d131160e4b2c0c3"
            );
        }
    }
}

pub(crate) mod sequence {
    pub(crate) const SEQUENCE_ACTION_BASE_FEE_KEY: &str = "seqbasefee";
    pub(crate) const SEQUENCE_ACTION_BYTE_COST_MULTIPLIER_KEY: &str = "seqmultiplier";
}

/// Helper struct whose `Display` impl outputs the prefix followed by the hex-encoded address.
pub(super) struct AddressPrefixer<'a, T> {
    prefix: &'static str,
    address: &'a T,
}

impl<'a, T> AddressPrefixer<'a, T> {
    pub(in crate::storage) fn new(prefix: &'static str, address: &'a T) -> Self {
        Self {
            prefix,
            address,
        }
    }
}

impl<'a, T: AddressBytes> Display for AddressPrefixer<'a, T> {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}{}",
            self.prefix,
            hex::encode(self.address.address_bytes())
        )
    }
}

/// Helper struct whose `Display` impl outputs the hex-encoded ibc-prefixed address, and that can be
/// parsed from such a hex-encoded form.
#[cfg_attr(test, derive(Debug, PartialEq))]
pub(super) struct Asset(IbcPrefixed);

impl Asset {
    pub(super) fn get(self) -> IbcPrefixed {
        self.0
    }
}

impl Display for Asset {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        f.write_str(&hex::encode(self.0.get()))
    }
}

impl<T: Into<IbcPrefixed>> From<T> for Asset {
    fn from(value: T) -> Self {
        Self(value.into())
    }
}

#[derive(Debug, thiserror::Error)]
#[error("failed to parse input as asset key")]
pub(super) struct ParseAssetKeyError {
    #[from]
    source: hex::FromHexError,
}

impl FromStr for Asset {
    type Err = ParseAssetKeyError;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        use hex::FromHex as _;
        let bytes = <[u8; 32]>::from_hex(s)?;
        Ok(Self(IbcPrefixed::new(bytes)))
    }
}

#[cfg(test)]
mod tests {
    use super::Asset;

    #[test]
    fn asset_key_to_string_parse_roundtrip() {
        let asset = "an/asset/with/a/prefix"
            .parse::<astria_core::primitive::v1::asset::Denom>()
            .unwrap();
        let expected = Asset::from(asset);
        let actual = expected.to_string().parse::<Asset>().unwrap();
        assert_eq!(expected, actual);
    }
}
