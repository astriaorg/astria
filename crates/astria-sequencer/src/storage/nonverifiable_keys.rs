pub(crate) mod api {
    use astria_core::primitive::v1::RollupId;

    pub(crate) fn block_hash_by_height_key(height: u64) -> Vec<u8> {
        [b"blockhash/".as_slice(), &height.to_le_bytes()].concat()
    }

    pub(crate) fn sequencer_block_header_by_hash_key(hash: &[u8]) -> Vec<u8> {
        [b"blockheader/", hash].concat()
    }

    pub(crate) fn rollup_data_by_hash_and_rollup_id_key(
        hash: &[u8],
        rollup_id: &RollupId,
    ) -> Vec<u8> {
        [b"rollupdata/", hash, rollup_id.as_ref()].concat()
    }

    pub(crate) fn rollup_ids_by_hash_key(hash: &[u8]) -> Vec<u8> {
        [b"rollupids/", hash].concat()
    }

    pub(crate) fn rollup_transactions_proof_by_hash_key(hash: &[u8]) -> Vec<u8> {
        [b"rolluptxsproof/", hash].concat()
    }

    pub(crate) fn rollup_ids_proof_by_hash_key(hash: &[u8]) -> Vec<u8> {
        [b"rollupidsproof/", hash].concat()
    }

    #[cfg(test)]
    mod tests {
        use telemetry::display::base64;

        use super::*;

        #[test]
        fn keys_should_not_change() {
            insta::assert_snapshot!(base64(&block_hash_by_height_key(42)));

            let hash = [1; 32];
            insta::assert_snapshot!(base64(&sequencer_block_header_by_hash_key(&hash)));

            let rollup_id = RollupId::new([2; 32]);
            insta::assert_snapshot!(base64(&rollup_data_by_hash_and_rollup_id_key(
                &hash, &rollup_id
            )));

            insta::assert_snapshot!(base64(&rollup_ids_by_hash_key(&hash)));

            insta::assert_snapshot!(base64(&rollup_transactions_proof_by_hash_key(&hash)));

            insta::assert_snapshot!(base64(&rollup_ids_proof_by_hash_key(&hash)));
        }
    }
}

pub(crate) mod app {
    pub(crate) fn storage_version_by_height_key(height: u64) -> Vec<u8> {
        format!("storage_version/{height}").into_bytes()
    }
}

pub(crate) mod assets {
    use astria_core::primitive::v1::asset::IbcPrefixed;
    use astria_eyre::eyre::{
        ContextCompat,
        Result,
        WrapErr,
    };

    use crate::storage::verifiable_keys::Asset;

    pub(crate) const BLOCK_FEES_PREFIX: &[u8] = b"block_fees/";
    pub(crate) const FEE_ASSET_PREFIX: &[u8] = b"fee_asset/";

    pub(crate) fn fee_asset_key<TAsset: Into<IbcPrefixed>>(asset: TAsset) -> Vec<u8> {
        [FEE_ASSET_PREFIX, Asset::from(asset).to_string().as_bytes()].concat()
    }

    pub(crate) fn block_fees_key<TAsset: Into<IbcPrefixed>>(asset: TAsset) -> Vec<u8> {
        [BLOCK_FEES_PREFIX, Asset::from(asset).to_string().as_bytes()].concat()
    }

    pub(crate) fn extract_asset_from_fee_asset_key(key: &[u8]) -> Result<IbcPrefixed> {
        extract_asset_from_key(key, FEE_ASSET_PREFIX)
            .wrap_err("failed to extract asset from fee asset key")
    }

    pub(crate) fn extract_asset_from_block_fees_key(key: &[u8]) -> Result<IbcPrefixed> {
        extract_asset_from_key(key, BLOCK_FEES_PREFIX)
            .wrap_err("failed to extract asset from fee asset key")
    }

    pub(crate) fn extract_asset_from_key(key: &[u8], prefix: &[u8]) -> Result<IbcPrefixed> {
        let suffix = key.strip_prefix(prefix).wrap_err_with(|| {
            format!(
                "key {} did not have prefix {}",
                telemetry::display::hex(key),
                telemetry::display::hex(prefix)
            )
        })?;
        let asset = std::str::from_utf8(suffix)
            .wrap_err("key suffix was not utf8 encoded; this should not happen")?
            .parse::<Asset>()
            .wrap_err("failed to parse storage key suffix as address hunk")?
            .get();
        Ok(asset)
    }

    #[cfg(test)]
    mod tests {
        use astria_core::primitive::v1::asset::Denom;

        use super::*;

        #[test]
        fn keys_should_not_change() {
            let trace_prefixed = "a/denom/with/a/prefix".parse::<Denom>().unwrap();
            insta::assert_snapshot!(std::str::from_utf8(&fee_asset_key(&trace_prefixed)).unwrap());
            insta::assert_snapshot!(std::str::from_utf8(&block_fees_key(&trace_prefixed)).unwrap());
        }

        #[test]
        fn should_extract_asset_from_key() {
            let asset = IbcPrefixed::new([1; 32]);

            let key = fee_asset_key(asset);
            let recovered_asset = extract_asset_from_fee_asset_key(&key).unwrap();
            assert_eq!(asset, recovered_asset);

            let key = block_fees_key(asset);
            let recovered_asset = extract_asset_from_block_fees_key(&key).unwrap();
            assert_eq!(asset, recovered_asset);
        }
    }
}

pub(crate) mod authority {
    pub(crate) const VALIDATOR_UPDATES_KEY: &[u8] = b"valupdates";
}

pub(crate) mod bridge {
    use astria_core::primitive::v1::RollupId;

    use crate::{
        accounts::AddressBytes,
        storage::verifiable_keys::{
            bridge::BRIDGE_ACCOUNT_PREFIX,
            AddressPrefixer,
        },
    };

    pub(crate) const DEPOSIT_PREFIX: &[u8] = b"deposit/";
    pub(crate) const DEPOSIT_NONCE_PREFIX: &[u8] = b"depositnonce/";

    pub(crate) fn deposit_key_prefix(rollup_id: &RollupId) -> Vec<u8> {
        [DEPOSIT_PREFIX, rollup_id.as_ref()].concat()
    }

    pub(crate) fn deposit_key(rollup_id: &RollupId, nonce: u32) -> Vec<u8> {
        [DEPOSIT_PREFIX, rollup_id.as_ref(), &nonce.to_le_bytes()].concat()
    }

    pub(crate) fn deposit_nonce_key(rollup_id: &RollupId) -> Vec<u8> {
        [DEPOSIT_NONCE_PREFIX, hex::encode(rollup_id).as_bytes()].concat()
    }

    pub(crate) fn last_transaction_id_for_bridge_account_key<T: AddressBytes>(
        address: &T,
    ) -> Vec<u8> {
        format!(
            "{}/lasttx",
            AddressPrefixer::new(BRIDGE_ACCOUNT_PREFIX, address)
        )
        .into_bytes()
    }

    #[cfg(test)]
    mod tests {
        use telemetry::display::base64;

        use super::*;

        #[test]
        fn keys_should_not_change() {
            let rollup_id = RollupId::new([1; 32]);
            insta::assert_snapshot!(base64(&deposit_key_prefix(&rollup_id)));
            insta::assert_snapshot!(base64(&deposit_key(&rollup_id, 42)));
            insta::assert_snapshot!(std::str::from_utf8(&deposit_nonce_key(&rollup_id)).unwrap());
            insta::assert_snapshot!(
                std::str::from_utf8(&last_transaction_id_for_bridge_account_key(&[2; 20])).unwrap()
            );
        }

        #[test]
        fn deposit_prefix_should_be_prefix_of_deposit_key() {
            let rollup_id = RollupId::new([1; 32]);
            let prefix = deposit_key_prefix(&rollup_id);
            let key = deposit_key(&rollup_id, 99);
            assert!(key.strip_prefix(prefix.as_slice()).is_some());
        }
    }
}
