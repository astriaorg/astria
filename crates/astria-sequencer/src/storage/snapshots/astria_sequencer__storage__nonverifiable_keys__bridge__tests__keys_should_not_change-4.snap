---
source: crates/astria-sequencer/src/storage/nonverifiable_keys.rs
assertion_line: 191
expression: "std::str::from_utf8(&last_transaction_id_for_bridge_account_key(&[2;\n20])).unwrap()"
---
bridgeacc/0202020202020202020202020202020202020202/lasttx
